module/zfs/zfs_vnops.c

```diff
@@ -3657,6 +3792,9 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 		error = zfs_link_create(tdl, szp, tx, ZRENAMING);
 		if (error == 0) {
 			szp->z_pflags |= ZFS_AV_MODIFIED;
+			if (tdzp->z_pflags & ZFS_PROJINHERIT &&
+			    S_ISDIR(ZTOI(szp)->i_mode))
+				szp->z_pflags |= ZFS_PROJINHERIT;
```

Comment:
In this case, we know that the szp has the same PROJID, but if PROJINHERIT is not already set on szp, how do we know if all the files/directories inside szp has the same PROJID?

I think we need to enforce that if tdzp has PROJINHERIT, then szp must *already* have PROJINHERIT set, so that we know all its descendents have the same PROJID.